File: leetcode/leetcode-patterns/minimum-depth-of-binary-tree.py
class Solution:
  def minDepth(self, root: TreeNode) -> int:
    if root is None:
      return 0
    
    depth = 1
    current_level = [root]
    while current_level:
      next_level = []
      for n in current_level:
        if not n.left and not n.right:
          return depth
        if n.left:
          next_level.append(n.left)
        if n.right:
          next_level.append(n.right)
      current_level = next_level
      depth += 1
    return depth

class Solution2:
  def minDepth(self, root: TreeNode) -> int:
    if root is None:
      return 0
    return self.inorder(root, 1)
  
  def inorder(self, node: TreeNode, depth: int) -> int:
    if node is None:
      return depth
    if node.left is None and node.right is None:
      return depth
    
    left_depth = right_depth = 100000
    if node.left:
      left_depth = self.inorder(node.left, depth+1)
    if node.right:
      right_depth = self.inorder(node.right, depth+1)
    
    return min(left_depth, right_depth)

File: leetcode/leetcode-patterns/minimum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function (root) {
  if (root === null) {
    return 0;
  }

  let currentLevel = [root];
  let depth = 1;
  while (currentLevel.length > 0) {
    let nextLevel = [];
    for (const node of currentLevel) {
      if (node.left === null && node.right === null) {
        return depth;
      }
      if (node.left) {
        nextLevel.push(node.left);
      }
      if (node.right) {
        nextLevel.push(node.right);
      }
    }
    depth++;
    currentLevel = nextLevel;
  }

  return depth;
};
